File: sol/solution/0101-0200/0103/Solution.php
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     function __construct($val = 0, $left = null, $right = null) {
 *         $this->val = $val;
 *         $this->left = $left;
 *         $this->right = $right;
 *     }
 * }
 */
class Solution {
    /**
     * @param TreeNode $root
     * @return Integer[][]
     */
    function zigzagLevelOrder($root) {
        $res = [];
        if (!$root) return $res;
        $queue = [$root];
        $leftToRight = true;
        while (!empty($queue)) {
            $level = [];
            $size = count($queue);
            for ($i = 0; $i < $size; $i++) {
                $node = array_shift($queue);
                if ($leftToRight) {
                    $level[] = $node->val;
                } else {
                    array_unshift($level, $node->val);
                }
                if ($node->left) $queue[] = $node->left;
                if ($node->right) $queue[] = $node->right;
            }
            $res[] = $level;
            $leftToRight = !$leftToRight;
        }
        return $res;
    }
}

File: sol/solution/0101-0200/0103/Solution.go
func zigzagLevelOrder(root *TreeNode) [][]int {
	if root == nil {
		return [][]int{}
	}
	var res [][]int
	queue := []*TreeNode{root}
	leftToRight := true
	for len(queue) > 0 {
		n := len(queue)
		level := make([]int, n)
		for i := 0; i < n; i++ {
			node := queue[0]
			queue = queue[1:]
			if leftToRight {
				level[i] = node.Val
			} else {
				level[n-1-i] = node.Val
			}
			if node.Left != nil {
				queue = append(queue, node.Left)
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
			}
		}
		res = append(res, level)
		leftToRight = !leftToRight
	}
	return res
}